_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+```properties
+# ...
+log.cleaner.io.max.bytes.per.second= 1.7976931348623157E308
+# ...
+```
+
+### Handling large message sizes
+
+Calibrated for optimal throughput in most situations, Kafka's default message batch size is set to 1MB. If you have the [disk capacity](#managing-disk-utilization), you can increase the batch size at a reduced throughput.  
+
+You have four ways to handle large message sizes:
+
+- Producer-side message compression
+- Reference-based messaging
+- Inline messaging
+- Broker and producer/consumer configuration
+
+We recommend trying producer-side message compression or reference-based messaging, which cover most situations.
+
+#### Producer-side compression
+For producer-side compression, you specify a `compression.type`, such as Gzip, to apply to batches. In the broker configuration, you use `compression.type=producer` so that the broker retains whatever compression the producer used. It's better that the producer and topic compression types match. Otherwise, the broker has to compress the batches again before appending them to the log.
+
+Compressing batches will add additional processing overhead on the producer and decompression overhead on the consumer. But you will get more data in a batch, which helps with throughput. Examine your metrics and combine producer-side compression with fine-tuning of the batch size to get the best outcome.
+
+#### Reference-based messaging
+Reference-based messaging only sends a reference to data stored in some other system in the message’s value. This is useful for data replication when you do not know how big a message will be.
+Data is written to the data store, which must be fast, durable, and highly available, and a reference to the data is returned. The producer sends the reference to Kafka. The consumer uses the reference to fetch the data from the data store.
+
+**Reference-based messaging flow**
+
+![Image of reference-based messaging flow](/assets/images/posts/2021-05-06-broker-tuning-reference-messaging.png)
+
+Referenced-based messaging means more for message passing requires more trips, so end-to-end latency will increase. One major drawback of this approach is that there is no automatic clean up of the data in the external system when the Kafka message is cleaned up.
+
+One way to mitigate the limitations of reference-based messaging is to take a hybrid approach. Send large messages to the data store and process standard-sized messages directly.
+
+#### Inline messaging
+Inline messaging is a complex process that splits messages into chunks that use the same key, which are then combined on output using a stream processor like Kafka Streams.
+
+The basic steps are:
+
+- The client application serializes then chunks the data if the message is too big.
```

Comment:
Let's try to be clear what is done for you by the producer and consumer v.s. what you have to do for yourself in the application logic. 

```suggestion
- The producing client application serializes then chunks the data if the message is too big.
```